actions/src/openstack_check_actions.py

```diff
@@ -0,0 +1,447 @@
+        Outputs a suitable dictionary to pass into create_tickets
+        """
+        output = {
+            "title": "Server {p[id]} has not been updated in more than 10 minutes during {p[action]}",
+            "body": "The following information may be useful\nHost id: {p[id]} \n{p[data]}",
+            "server_list": [],
+        }
+
+        if all_projects:
+            with OpenstackConnection(cloud_name=cloud_account) as conn:
+                projects = conn.list_projects()
+            for project in projects:
+                deleted = self.check_deleted(project=project["id"], cloud=cloud_account)
+            output["server_list"].extend(deleted)
+        else:
+            deleted = self.check_deleted(project=project_id, cloud=cloud_account)
+            output["server_list"] = deleted
+
+        return output
+
+    def check_deleted(self, cloud: str, project: str):
+        """
+        Runs the check for deleting machines on a per-project basis
+        """
+        output = []
+
+        print("Checking project", project)
+        with OpenstackConnection(cloud_name=cloud) as conn:
+            server_list = conn.list_servers(
+                filters={"all_tenants": True, "project_id": project}
+            )
+        # Loop through each server in the project/list
+        for i in server_list:
+            # Take current time and check difference between updated time
+            since_update = datetime.utcnow() - datetime.strptime(
+                i["updated"], "%Y-%m-%dT%H:%M:%Sz"
+            )
+            # Check if server has been stuck in deleting for too long.
+            # (uses the last updated time so if changes have been made
+            # to the server while deleting the check may not work.)
+            if i["status"] == "deleting" and since_update.total_seconds() >= 600:
+                # Append data to output array
+                output.append(
+                    {
+                        "dataTitle": {"id": str(i.id), "action": str(i["status"])},
+                        "dataBody": {"id": i["id"], "data": json.dumps(i)},
+                    }
+                )
+
+        return output
+
+    def loadbalancer_check(self, cloud_account: str):
+        """
+        Action to check loadbalancer and amphorae status
+        output suitable to be passed to create_tickets
+        """
+
+        with OpenstackConnection(cloud_name=cloud_account) as conn:
+            if cloud_account == "dev":
+                amphorae = requests.get(
+                    "https://dev-openstack.nubes.rl.ac.uk:9876/v2/octavia/amphorae",
+                    headers={"X-Auth-Token": conn.auth_token},
+                )
+            else:
+                amphorae = requests.get(
+                    "https://openstack.nubes.rl.ac.uk:9876/v2/octavia/amphorae",
+                    headers={"X-Auth-Token": conn.auth_token()},
+                )
```

Comment:
Can we move Amphorae handling into a separate file as we can reuse these chunks in future code (and it helps keep everything tidy). It could go into the `lib/` folder in a new `amphorae` module

Also this approach of checking for `dev` is quite brittle, but I can't think of a better approach.
I wonder if there's a way to get the Openstack API to effectively handle this from the URL in the clouds.yaml file :thinking:

Reply:
Missed this in the latest commit. I'll move it around now.